File: Remote/Source/Networking/MessageQueueEntry.m
//
//  MessageQueueEntry.m
//  Remote
//
//  Created by Jason Cardwell on 9/8/14.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

#import "MessageQueueEntry.h"
@import Moonkit;

@implementation MessageQueueEntry

/// message:completion:
/// @param message
/// @param completion
/// @return instancetype
+ (instancetype)message:(NSString *)message
             completion:(void (^)(BOOL success, NSString * response, NSError * error))completion
{

  if (StringIsEmpty(message)) ThrowInvalidNilArgument(message);

  MessageQueueEntry * entry = [self new];
  entry.message    = message;
  entry.completion = completion;
  entry.userInfo   = [MSDictionary dictionary];

  return entry;
  
}

/// data
/// @return NSData *
- (NSData *)data { return (self.message ? [self.message dataUsingEncoding:NSUTF8StringEncoding] : nil); }

/// description
/// @return NSString
- (NSString *)description {

  return $(@"MessageQueueEntry(\n"
           "  message: '%@'\n"
           "  userInfo: {\n"
           "    %@\n"
           "  }", [self.message stringByReplacingReturnsWithSymbol], [self.userInfo formattedDescription]);

}

@end

File: Networking/Source/MessageQueueEntry.swift
//
//  MessageQueueEntry.swift
//  Remote
//
//  Created by Jason Cardwell on 5/05/15.
//  Copyright (c) 2014 Moondeer Studios. All rights reserved.
//

import Foundation
import MoonKit

protocol MessageData {
  var msg: String { get }
  var data: NSData { get }
}


struct MessageQueueEntry<T:MessageData> {

  typealias Callback = (Bool, ErrorType?) -> Void

  var messageData: T

  /** The message being sent */
  var message: String { return messageData.msg }

  /** The message being sent as data */
  var data: NSData { return messageData.data }

  /** Generalized storage space */
  var info: [String:Any]

  /** Optional callback to invoke after a corresponding response has been received for the message */
  let completion: Callback?

  /**
  initWithMessageData:info:completion:

  - parameter d: T
  - parameter i: [String:Any]
  - parameter c: Callback? = nil
  */
  init(messageData d: T, info i: [String:Any] = [:], completion c: Callback? = nil) {
    messageData = d
    info = i
    completion = c
  }

}